File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.js
function solution(n, costs) {
    // Union-Find
    const parent = Array(n);
    for (let i = 0; i < n; i++) {
        parent[i] = i;
    }

    function find(x) {
        if (parent[x] !== x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    function union(x, y) {
        const rx = find(x);
        const ry = find(y);
        if (rx !== ry) {
            parent[rx] = ry;
        }
    }

    // 1. 비용 오름차순 정렬
    costs.sort((a, b) => a[2] - b[2]);

    let answer = 0;
    let edges = 0;

    for (const [from, to, cost] of costs) {
        if (find(from) !== find(to)) {
            union(from, to);
            answer += cost;
            edges++;
            if (edges === n - 1) break;
        }
    }

    return answer;
}

File: 프로그래머스/3/42861. 섬 연결하기/섬 연결하기.java
import java.util.*;

class Solution {
    int[] parent;

    public int solution(int n, int[][] costs) {
        parent = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
        }

        // 1. 비용 오름차순 정렬
        Arrays.sort(costs, (c1, c2) -> c1[2] - c2[2]);

        int answer = 0;
        int edges = 0;

        for (int[] edge : costs) {
            int a = edge[0];
            int b = edge[1];
            int cost = edge[2];

            if (find(a) != find(b)) {
                union(a, b);
                answer += cost;
                edges++;
                if (edges == n - 1) break;  // MST 완성
            }
        }

        return answer;
    }

    private int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]);
        }
        return parent[x];
    }

    private void union(int x, int y) {
        int rx = find(x);
        int ry = find(y);
        if (rx != ry) {
            parent[rx] = ry;
        }
    }
}